### ISSUE TITLE
Plan for high level graph optimizations

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#5644]
Now that we have high level graphs in our collections, we can do some more complex optimizations.  We've done a little bit of this with blockwise fusion, read_parquet and column selection, and root fusion, all of which have had really positive effects. 

There is likely more that we can do here.  Some thoughts ...

-  Fuse column selection with other kinds of data access, like read_csv or ORC
-  Pass slicing through some blockwise operations in Dask array (a long-held request by @shoyer)
-  Optimize the subgraph callables in blockwise with Numba (or something else) for dask array to avoid memory copies and maybe reduce serialization time
-  Swap around Joins and filters and column access
-  ...

It would be nice to have a current maintainer consider these options, guess how long they would take, and see what makes sense to do short term

**Comment by simone-codeluppi** [comment]
So if I understand correctly the `dask.optimization` utilities do not currently works on HighLevelGraphs, correct? I am curious because I would like to fuse some tasks in parallel branches of my processing graph generated using delayed functions to make sure that all the step in a branch will run on the same worker/node.

**Comment by TomAugspurger** [comment]
> dask.optimization utilities do not currently works on HighLevelGraphs, correct?

No, they do. See https://github.com/dask/dask/blob/1c39d48b452d857418c854008932fdac2608e121/dask/array/optimization.py#L40-L43 for example.

> I would like to fuse some tasks in parallel branches of my processing graph generated using delayed functions 

I'm not sure if this is likely to get fused. HighLevelGraphs rely on the structure provided by Dask's collections, and `dask.delayed` has less structure. But it's worth trying out.

Keep in mind that Dask's low-level fusion in https://github.com/dask/dask/blob/1c39d48b452d857418c854008932fdac2608e121/dask/optimization.py#L401-L411 may also kick in.

**Comment by TomAugspurger** [comment]
A slightly more specific optimization is pushing down binary operations and masking on DataFrames read from parquet files into the `filters` argument.

```python
>>> df = dd.read_parquet("data.parquet")
>>> df[df.A > 1] 
```

Would become something like

```python
>>> dd.read_parquet("data.parquet", filters=[("A",  ">", 1")])
```

There are probably lots of complexity, but that might make a good first introduction to this (cc @eriknw)

**Comment by chenliu0831** [comment]
Just curious and a quick imagination, does it make sense to compile the graph to some IR in the compiler world and leverage LLVM or other compiler technologies to handle graph optimizations (like convert the graph to SSA form)?

**Comment by jcrist** [comment]
> Just curious and a quick imagination, does it make sense to compile the graph to some IR in the compiler world and leverage LLVM or other compiler technologies to handle graph optimizations (like convert the graph to SSA form)?

Not really, compiler IR is generally lower level than what we describe in dask graphs, the transformations we want to do and the transformations compilers want to do are sufficiently different that we'd get no benefit (AFAICT) from doing so.

**Comment by jcrist** [comment]
Was just helping @jsignell debug an issue that had to do with our optimization passes, and the best solution I could come up with is some change in how we manage collection-specific optimizations.

Currently, different collections have different `__dask_optimize__` methods with the following signature:

```python
def __dask_optimize__(self, dsk: Mapping, keys: Iterable, **kwargs) -> dict:
    ...
```

Depending on the call path `dsk` may be either a `HighLevelGraph` or a `dict`. If provided a `HighLevelGraph`, these may do some high-level optimizations, before lowering to a `dict` based graph and doing things like `cull` and `fuse`. See the one for `dask.dataframe` for example:

https://github.com/dask/dask/blob/8eeb0e0194ef0561b4202f42de06c0b7fc0784b9/dask/dataframe/optimize.py#L12-L35

When converting between collections `A` to `B` we usually (but not always) call the optimization function of `A` on the graph to ensure that its optimizations are run, since `B` will no longer track those optimizations.

This can cause issues, especially if `fuse` is enabled. For example, say we have 2 `dask.dataframe` objects that share some upstream graph structure:

```
df1, df2 = some_shared_upstream_task()
a = df1.to_delayed()
b = df2.to_delayed()
```

If fusion is enabled, there is a risk that the call to `__dask_optimize__` inside `to_delayed` will fuse out the shared upstream tasks, resulting in duplicated work (see #6232).

A few possible solutions:

1. Since many collections share the same low-level graph optimizations - `cull`, and `fuse` - we may only need the high-level optimizations to be class-specific. This currently isn't 100% true - `dask.bag` has some class-specific optimization passes that are implemented on the low-level graphs, but with some work we may be able to change that around. Then `__dask_optimize__` would map `HighLevelGraph -> HighLevelGraph`, and there'd be a consistent low-level optimization pipeline on the merged graph handling the `cull`, `fuse`, etc...

2. We might instead associate optimization functions with layers in the `HighLevelGraph` instead of with collections. Then the optimizations wouldn't need to run on conversion between collection types, the backing layers could be merged and optimized on compute. I suspect this would be harder to implement than option 1, but may solve some cases not solved that solution.

There also might be a simpler thing I didn't think of, a bit in the weeds here.

**Comment by jsignell** [comment]
Thanks for writing that up Jim. I think option 1 sounds more straightforward and would avoid adding more information to the `HighLevelGraph` object (something that I am newly wary of bloating).

**Comment by TomAugspurger** [comment]
In https://github.com/dask/dask/pull/6262 @gforsyth added a CSVSubgraph, which makes the structure of `read_csv` similar to `read_parquet`. We'll now be able to implement an optimization similar to `optimize_read_parquet_getitem`, but for `read_csv` instead. Finding the patterns that can be shared between the two optimizations will be an interesting exercise.

**Comment by mrocklin** [comment]
> Finding the patterns that can be shared between the two optimizations
will be an interesting exercise.

+1

**Comment by nils-braun** [comment]
Recently, @beckernick brought my attention to https://github.com/IntelPython/sdc (in https://github.com/nils-braun/dask-sql/issues/143). This project offers numba optimizations for pandas operations. In my very naive world I would think this (and numba itself) can be used on fused subgraphs to compile the code before distributing it to the workers (so compile it once and run it multiple times) - similar to what @jcrist started in https://github.com/dask/dask/pull/4091/files as a first draft (but for all sorts of fused subgraphs). 

I know that this is so far only a very rough draft, but I would imagine it to work similarly to this:
1. While building the highlevel graph using some collection type, like dataframes, we already know the type information due to the _meta. We can store this "somehow" (with annotations https://github.com/dask/dask/issues/3783?) in the individual tasks or layers (currently, this is "lost" when moving from collections to graphs - or?). This would mean not every task will get a _meta annotation (as we only know the _meta of collections), but at least a few of them. 
2. When fusing subgraphs, we use the type annotations from the inputs (if we have it) and create a pre-compiled version of the subgraph function. This can replace the original function in the graph.
3. The compiled function can (hopefully) be executed on the workers using the real inputs. We can always fall back to the non-compiled version, e.g. if the meta type information was wrong.

I do not have any feeling if the compilation overhead (and implementation work!) is worth the gains, but I just know that many other distributed engines (Spark, Impala, ...) use code generation (so there must be at least anything to gain here...). 

Before one would go into more details on this, there are a couple of questions to answer:
* is the _meta information enough to pre-compile the fused function?
* how to properly preserve the meta information from the collection to the dask graph during rewriting, optimization, etc.? where should this happen?
* was this already tried (apart from https://github.com/dask/dask/pull/4091/files)? Or was it deemed infeasible already? (sorry if I just have not found it)
